channeldb/control_tower.go

```diff
@@ -401,6 +487,129 @@ type InFlightPayment struct {
 	Attempt *PaymentAttemptInfo
 }
 
+// SubscribePayment subscribes to updates for the payment with the given hash.
+// It returns a boolean indicating whether the payment is still in flight and a
+// channel that provides the final outcome of the payment.
+func (p *paymentControl) SubscribePayment(paymentHash lntypes.Hash) (
+	bool, chan PaymentEvent, error) {
+
+	// Create a channel with buffer size 1. For every payment there will be
+	// exactly one event sent.
+	c := make(chan PaymentEvent, 1)
+
+	// Lock the hash to prevent a race condition with subscribing and an
+	// update from router.
+	p.hashMutex.Lock(paymentHash)
+	defer p.hashMutex.Unlock(paymentHash)
+
+	var inFlight bool
+
+	err := p.db.View(func(tx *bbolt.Tx) error {
+		bucket, err := fetchPaymentBucket(tx, paymentHash)
+		if err != nil {
+			return err
+		}
+
+		var event PaymentEvent
+
+		status := fetchPaymentStatus(bucket)
+		switch status {
+
+		// Payment is currently in flight. Register this subscriber and
+		// return without writing a result to the channel yet.
+		case StatusInFlight:
+			p.subscribersMtx.Lock()
```

Comment:
Grabbing mutexes in a database transaction makes be uneasy...

Alternatively, I think we can lift most of this logic _outside_ of the transaction by optimistically fetching everything we _might_ need. Returning back to my prior comment, if the subscription handling wasn't so coupled to the underlying database logic, there would be no concern here.

Reply:
Addressed by splitting off subscriber logic.